display_list/geometry/dl_region.cc

```diff
@@ -60,189 +626,116 @@ std::vector<SkIRect> DlRegion::getRects(bool deband) const {
-      continue;
-    }
-    size_t last_index = i;
-    while (last_index + 1 < size && right >= spans[last_index + 1].left) {
-      ++last_index;
-    }
-    span.left = std::min(span.left, left);
-    span.right = std::max(spans[last_index].right, right);
-    if (last_index > i) {
-      spans.erase(spans.begin() + i + 1, spans.begin() + last_index + 1);
-    }
-    return;
-  }
+DlRegion::~DlRegion() {}
 
-  spans.push_back({left, right});
+bool DlRegion::isComplex() const {
+  return lines_.size() > 1 ||
+         (lines_.size() == 1 &&
+          span_buffer_.getChunkSize(lines_.front().chunk_handle) > 1);
 }
 
-bool DlRegion::SpanLine::spansEqual(const SpanLine& l2) const {
-  SpanVec& spans = *this->spans;
-  SpanVec& otherSpans = *l2.spans;
-  FML_DCHECK(this != &l2);
-
-  if (spans.size() != otherSpans.size()) {
+bool DlRegion::intersects(const SkIRect& rect) const {
+  if (isEmpty()) {
     return false;
   }
-  return memcmp(spans.data(), otherSpans.data(), spans.size() * sizeof(Span)) ==
-         0;
-}
-
-void DlRegion::insertLine(size_t position, SpanLine line) {
-  lines_.insert(lines_.begin() + position, line);
-}
 
-DlRegion::LineVec::iterator DlRegion::removeLine(
-    DlRegion::LineVec::iterator line) {
-  spanvec_pool_.push_back(line->spans);
-  return lines_.erase(line);
-}
+  auto bounds_intersect = SkIRect::Intersects(bounds_, rect);
 
-DlRegion::SpanLine DlRegion::makeLine(int32_t top,
-                                      int32_t bottom,
-                                      int32_t spanLeft,
-                                      int32_t spanRight) {
-  SpanVec* span_vec;
-  if (!spanvec_pool_.empty()) {
-    span_vec = spanvec_pool_.back();
-    spanvec_pool_.pop_back();
-    span_vec->clear();
-  } else {
-    span_vec = new SpanVec();
+  if (isSimple()) {
+    return bounds_intersect;
   }
-  span_vec->push_back({spanLeft, spanRight});
-  return {top, bottom, span_vec};
-}
 
-DlRegion::SpanLine DlRegion::makeLine(int32_t top,
-                                      int32_t bottom,
-                                      const SpanVec& spans) {
-  SpanVec* span_vec;
-  if (!spanvec_pool_.empty()) {
-    span_vec = spanvec_pool_.back();
-    spanvec_pool_.pop_back();
-  } else {
-    span_vec = new SpanVec();
+  if (!bounds_intersect) {
+    return false;
   }
-  *span_vec = spans;
-  return {top, bottom, span_vec};
-}
-
-void DlRegion::addRects(std::vector<SkIRect>&& rects) {
-  std::sort(rects.begin(), rects.end(), [](const SkIRect& a, const SkIRect& b) {
-    // Sort the rectangles by Y axis. Because the rectangles have varying
-    // height, they are added to span lines in non-deterministic order and thus
-    // it makes no difference if they are also sorted by the X axis.
-    return a.top() < b.top();
-  });
 
-  size_t start_index = 0;
+  auto it = std::upper_bound(
+      lines_.begin(), lines_.end(), rect.fTop,
+      [](int32_t i, const SpanLine& line) { return i < line.bottom; });
```

Comment:
i => top?

Reply:
I think the condition is correct, but I'll rewrite it with `lower_bound` which should make it clearer.

Reply:
Oh, I'm sorry, my comment was confusing.

The condition and use of upper_bound was correct and I wasn't referring to that. I was referring to the variable naming. In other places you compare the lines to a variable named "top" for clarity. "=>" meant "change the name(?)", not a backwards comparison...

Reply:
🤦Of course :) Updated.